### PR TITLE
Jetty 9.4.x http interceptor #382

### PR DESCRIPTION
**Pull request by gregw**
Implementation for #382  HTTP Input interception with a GZIP inflator usage.

Creating PR for review.

### CODE REVIEW

File: jetty-client/src/main/java/org/eclipse/jetty/client/GZIPContentDecoder.java [hunk 1]
```diff
@@ -18,26 +18,13 @@
...
-import java.util.zip.ZipException;
-
-import org.eclipse.jetty.util.BufferUtil;
 
 /**
  * {@link ContentDecoder} for the "gzip" encoding.
+ * 
  */
-public class GZIPContentDecoder implements ContentDecoder
+public class GZIPContentDecoder extends org.eclipse.jetty.http.GZIPContentDecoder implements ContentDecoder
```
Comment (gregw) [inline jetty-client/src/main/java/org/eclipse/jetty/client/GZIPContentDecoder.java]: This now uses the common implementation from HTTP indirectly.   Perhaps the client could be changed to use it directly?

File: jetty-client/src/test/java/org/eclipse/jetty/client/GZIPContentDecoderTest.java [hunk 2]
```diff
@@ -33,6 +33,7 @@
 import org.junit.Rule;
 import org.junit.Test;
 
+@Deprecated
```
Comment (gregw) [inline jetty-client/src/test/java/org/eclipse/jetty/client/GZIPContentDecoderTest.java]: Don't need this test long term as it is duplicated in jetty-http

File: jetty-http/src/main/java/org/eclipse/jetty/http/GZIPContentDecoder.java [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+     * Derived implementations may choose to consume chunks individually
+     * and return false to prevent further inflation until a subsequent
+     * call to {@link #decode(ByteBuffer)} or {@link #decodeChunks(ByteBuffer)}.
+     *  
+     * @param chunk The inflated chunk of data
+     * @return False if inflating should continue, or True if the call 
+     * to {@link #decodeChunks(ByteBuffer)} or {@link #decode(ByteBuffer)} 
+     * should return, allowing back pressure of compressed data.
+     */
+    protected boolean decodedChunk(ByteBuffer chunk)
```
Comment (gregw) [inline jetty-http/src/main/java/org/eclipse/jetty/http/GZIPContentDecoder.java]: This method allows back pressure during inflation by returning true.  Previous impl always fully inflated.... perhaps the client could also use this?

File: jetty-http/src/main/java/org/eclipse/jetty/http/GZIPContentDecoder.java [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+     * more compressed data or a call to {@link #decodedChunk(ByteBuffer)} returns true.
+     * @param compressed
+     */
+    protected void decodeChunks(ByteBuffer compressed)
+    {
+        try
+        {
+            while (true)
+            {
+                switch (_state)
```
Comment (gregw) [inline jetty-http/src/main/java/org/eclipse/jetty/http/GZIPContentDecoder.java]: Split the state machine into states that consume byte by byte and those that do not.

File: jetty-server/src/main/java/org/eclipse/jetty/server/HttpInput.java [hunk 5]
```diff
@@ -223,73 +308,119 @@ protected Content nextContent() throws IOException
...
+        
     }
 
     /**
+     * Poll the inputQ for Content or EOF. Consumed buffers and non EOF {@link PoisonPillContent}s are removed. EOF state is not updated.
+     * Interception is done within this method.
+     * @return Content, EOF or null
      */
-    protected void consumeNonContent()
+    protected Content pollReadable()
```
Comment (gregw) [inline jetty-server/src/main/java/org/eclipse/jetty/server/HttpInput.java]: reduced duplicate code from pollContent into this method

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipHandler.java [hunk 6]
```diff
@@ -400,6 +402,24 @@ protected HttpField getVaryField()
 
     /* ------------------------------------------------------------ */
     /**
+     * @return size in bytes of the buffer to inflate compressed request, or 0 for no inflation.
```
Comment (gregw) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipHandler.java]: Should request inflation have its own Handler?   maybe but I think this is not bad either.

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipHandler.java [hunk 7]
```diff
@@ -409,6 +429,19 @@ public void handle(String target, Request baseRequest, HttpServletRequest reques
         String path = context==null?baseRequest.getRequestURI():URIUtil.addPaths(baseRequest.getServletPath(),baseRequest.getPathInfo());
         LOG.debug("{} handle {} in {}",this,baseRequest,context);
         
+        // Handle request inflation
+        if (_inflateBufferSize>0)
+        {
+            HttpField ce = baseRequest.getHttpFields().getField(HttpHeader.CONTENT_ENCODING);
+            if (ce!=null && "gzip".equalsIgnoreCase(ce.getValue()))
+            {
+                // TODO should check ce.contains and then remove just the gzip encoding
```
Comment (gregw) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipHandler.java]: Do we need to handle multiple encodings?   No for now.

**Comment by sbordet** [comment]
@gregw, overall I think we should consider a different, non-blocking design for the input interceptor.
Also, your work on backpressure revealed an issue on the client with a gzip bomb, so I think the client definitely needs to use this new class with its primitives, rather than just calling `decode()`.

**Comment by gregw** [comment]
@sbordet The interceptor is non blocking.  readFrom should definitely not block and should only take content from another content - which if the source content is empty, then the empty content is returned.

So I think the design is non blocking... perhaps I need to be clearer in the javadoc and/or naming?

**Comment by sbordet** [comment]
So, for example, how can I translate the content from English to Italian ?
I get the content in `readFrom(ByteBuffer content)`, then I must call an external service to translate. What do I return meanwhile ? An empty buffer ? But `content` will be consumed (I sent it to the translation service), so the code cannot assume that just because `content` is consumed then what is returned (the empty buffer) is the intercepted content.
Otherwise I would need to copy `content`, which is not optimal.
We actually need 2 results: the indication that the interception is finished (now conveyed by `content` being consumed) and the N intercepted buffers, where N >= 1.

**Comment by gregw** [comment]
I see that use-case, but I don't think it really belongs in the HttpInput Interceptor.  That is the use-case for the middleman servlet - or a reactive streams framework.    The HttpInput Interceptor is symmetric to the existing HttpOutput.Interceptor, which also does not support such async implementations.

**Comment by sbordet** [comment]
@gregw sure `HttpOutput.Interceptor` supports non-blocking writes, because it takes a `Callback`.
What am I missing ?

And I think the interceptors are indeed for content transformation, gzip being the primary example.
I don't see how gzip is different from translating English to Italian, or JSON to XML or rewrites, etc.

That is what you would do with a Servlet Filter wrapping, except it does not support non-blocking.

What am I missing ?

**Comment by gregw** [comment]
@sbordet well I'm convinced.... but for another reason.  Because an interceptor can fully consume content, we have to intercept before we decide to call onDataAvailable for an async reader, otherwise something written using read() may fail if zero bytes are available.

Once we involve that scheduling decision, we can then implement your translating interceptor:
- it initially consumes all the data (and sends it to translate service)
- eventually translate service responds with translation, which is put into a "poison pill" Content that looks to be zero length and HttpInput.addContent is called.
- the poison pill arrives at the interceptor and is inflated into the translated content and then onDataAvailable is called

**Comment by gregw** [comment]
The translate algorithm I posted above does not work as it allows the EOF content to overtake translated content.

**Pull request merged**